### PR TITLE
Bugs/9483 - fixes SSO signon for JobScheduler AND  Chromium OIDC correlation issues.

### PR DESCRIPTION
**Pull request by bradhead**
# Fixes or Implements [AB#9483](https://qslvic.visualstudio.com/304a1f8c-dace-4f85-adf3-bf563d5b3a39/_workitems/edit/9483)

* [ ] Enhancement
* [x ] Bug
* [ ] Documentation

## Description

Refactored Hangfire to use modern OIDC auth2 in middleware.  
Added checks to relax SameSite and CookiePolicy, Nonce for when developing on localhost on Chromium, Edge without HTTPS.
Fixes deprecated time in JobScheduler email jobs.
Moves OIDC Authentication Services Configuration into Common Code - to be also used by AdminWeb in future refactoring.
Fixes a few non-related code smells.

## Testing

* [ ] Unit Tests Updated
* [ ] Functional Tests Updated
* [x] Not Required

### Steps to Reproduce

Long standing OIDC SSO login issue with Hangfire. 

### UI Changes

YES | NO

### Browsers Tested

* [x] Chrome
* [x] Safari
* [x] Edge
* [ ] Firefox

Provide an explanation for any test(s) not completed.

## Notes/Additional Comments

Any additional notes or comments that may be relevant.  Include any specialized deployment steps here.  

</br>
</br>
</br>
___
  
## Reviewer Guidelines

A reviewer should examine the PR if the Status above indicates the PR is **Ready**.
Once ready, the reviewer should read the information below as well as supporting information like the Azure ticket to ensure they have a good understanding of the request.

The reviewer should then:

1. Is the PR template complete and accurate?

    Confirm by reading all sections below.

1. Confirm the merge branch

    A normal feature/bug branch should merge into the Dev branch.  Exceptions for hotfix and sprint merges do occur and the target would then be Master.

1. Do the code commits have an appropriate comment and include linkage to the azure ticket by including a reference to "AB#nnnn"?

1. Review the Code Changes

    1. Do the changes match the description?

        The code commits should reflect the documented description and nothing more.  If additional changes are included the reviewer should call these out and request that the developer document the changes in Azure and in the description or remove the changed code.

    1. useful tips

        * Any obvious errors or omissions?
        * Does the code match the style of other similar work?
        * Is the code easy to understand?
        * Is the code well documented?
        * DB Migrations are appropriate?

If the Answer to any of the questions above is No, then a request for change should be the result of the review.

### CODE REVIEW

File: Apps/AdminWebClient/README.md [hunk 1]
```diff
@@ -39,7 +39,7 @@ dotnet run
 Open another terminal Window to run the VUE Cli
 
 ```bash
-cd $GATEWAYHOME/Apps/AdminWebClient/src/
+cd $GATEWAYHOME/Apps/AdminWebClient/src/WebClient
```
Comment (sslaws) [inline Apps/AdminWebClient/README.md:42]: This should be ClientApp instead of WebClient

File: Apps/Common/src/AspNetConfiguration/StartupConfiguration.cs [hunk 2]
```diff
@@ -509,9 +608,11 @@ public void UseHttp(IApplicationBuilder app)
             else
             {
                 // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
+                this.Logger.LogInformation("Using HSTS, which sets Strict-Transport-Security Header");
```
Comment (sslaws) [inline Apps/Common/src/AspNetConfiguration/StartupConfiguration.cs:611]: Why did you remove this logging?  It only prints on startup?

**Review by sslaws — CHANGES_REQUESTED** [review]

File: Apps/Common/src/AspNetConfiguration/StartupConfiguration.cs (hunk 2 above)
Comment (FuriousLlama) [reply Apps/Common/src/AspNetConfiguration/StartupConfiguration.cs:611]: looks to me that it was added instead of removed

File: Apps/Common/src/AspNetConfiguration/StartupConfiguration.cs [hunk 3]
```diff
@@ -307,6 +312,100 @@ public void ConfigureAuthServicesForJwtBearer(IServiceCollection services)
             });
         }
 
+        /// <summary>
+        /// This sets up the OIDC authentication.
+        /// </summary>
+        /// <param name="services">The passed in IServiceCollection.</param>
+        public void ConfigureOpenIdConnectServices(IServiceCollection services)
```
Comment (FuriousLlama) [inline Apps/Common/src/AspNetConfiguration/StartupConfiguration.cs:319]: this method seems like is only used in Jobscheduler.  and has hardcoded cookie names for said project. If not used anywhere else, lets move it out of common

File: Apps/AdminWebClient/README.md (hunk 1 above)
Comment (bradhead) [reply Apps/AdminWebClient/README.md:42]: Ah yes. will fix.

File: Apps/Common/src/AspNetConfiguration/StartupConfiguration.cs (hunk 3 above)
Comment (bradhead) [reply Apps/Common/src/AspNetConfiguration/StartupConfiguration.cs:319]: Forgot to refactor that.  The plan is to use this common across JobScheduler and AdminWebClient. Will refactor.

**Review by sslaws — APPROVED** [review]

File: Apps/Common/src/AspNetConfiguration/StartupConfiguration.cs [hunk 4]
```diff
@@ -509,9 +608,16 @@ public void UseHttp(IApplicationBuilder app)
...
+                this.Logger.LogInformation("Using HSTS, which sets Strict-Transport-Security Header");
                 app.UseHsts();
             }
 
+            if (!this.environment.IsDevelopment())
+            {
+                app.UseHttpsRedirection();
+            }
+
+            app.UseStaticFiles();
```
Comment (FuriousLlama) [inline Apps/Common/src/AspNetConfiguration/StartupConfiguration.cs:620]: I dont think use static files is necessary for this method since its used on the applications that dont serve static content

Comment (bradhead) [reply Apps/Common/src/AspNetConfiguration/StartupConfiguration.cs:620]: I didn't add that, it was there before. AdminWebClient may use this common code and it does have static files. It doesn't seem to have any negative affect on Hangfire.

Comment (bradhead) [reply Apps/Common/src/AspNetConfiguration/StartupConfiguration.cs:620]: I had removed it, then added it back. The main thing I added was to not use Https redirection when in Development mode.

Comment (FuriousLlama) [reply Apps/Common/src/AspNetConfiguration/StartupConfiguration.cs:620]: thats what I saw. UseStaticFiles is used in Admin/Startup. I think we can remove it here

Comment (bradhead) [reply Apps/Common/src/AspNetConfiguration/StartupConfiguration.cs:620]: ok sounds good. I just worry about ordering sensitivity as this seems to be fussy on ordering, at least services configuration sure is!

Comment (bradhead) [reply Apps/Common/src/AspNetConfiguration/StartupConfiguration.cs:620]: In particular everywhere I look the app.Use* ordering is the same, with UseStaticFiles coming before UseRouting. I worry that this order if messed up will create side effects.

**Review by FuriousLlama — APPROVED** [review]

File: Apps/Common/src/AspNetConfiguration/StartupConfiguration.cs (hunk 4 above)
Comment (bradhead) [reply Apps/Common/src/AspNetConfiguration/StartupConfiguration.cs:620]: i.e. https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-3.1#middleware-order
